src/shared/components/stepper/StepperContext.tsx

```diff
@@ -0,0 +1,141 @@
+import { createContext, FC, useCallback, useReducer } from "react";
+
+interface StepperContextStateShape {
+  labels: string[];
+  currentStep: number;
+  next: () => void;
+  back: () => void;
+  toStep: (step: number) => void;
+  completed: Record<number, boolean>;
+}
+
+const stepperContextInitialState = () => ({
+  currentStep: 0,
+  next: () => {},
+  back: () => {},
+  toStep: () => {},
+  completed: {},
+  labels: [],
+});
+
+interface StepperInternalStateShape {
+  steps: number[];
+  currentStep: number;
+  completed: Record<number, boolean>;
+}
+
+const stepperInternalInitialState = (
+  labels: string[]
+): StepperInternalStateShape => ({
+  steps: getStepsArray(labels.length),
+  currentStep: 0,
+  completed: {},
+});
+
+enum StepperInternalActionTypes {
+  next = "STEPPER/next",
+  back = "STEPPER/back",
+  toStep = "STEPPER/toStep",
+}
+
+type StepperInternalAction =
+  | { type: StepperInternalActionTypes.next }
+  | { type: StepperInternalActionTypes.back }
+  | { type: StepperInternalActionTypes.toStep; payload: { step: number } };
+
+const next = (): StepperInternalAction => ({
+  type: StepperInternalActionTypes.next,
+});
+
+const back = (): StepperInternalAction => ({
+  type: StepperInternalActionTypes.back,
+});
+
+const toStep = (step: number): StepperInternalAction => ({
+  type: StepperInternalActionTypes.toStep,
+  payload: { step },
+});
+
+const withinRange = (min: number, max: number, value: number) =>
+  value >= min && value <= max;
+
+const reducer = (
+  state: StepperInternalStateShape,
+  action: StepperInternalAction
+) => {
+  switch (action.type) {
+    case StepperInternalActionTypes.next: {
+      const { steps, completed } = state;
+      const { currentStep } = state;
+
+      const nextStep = currentStep + 1;
+      if (withinRange(0, steps.length, nextStep)) {
+        return {
+          ...state,
+          currentStep: currentStep + 1,
+          completed: { ...completed, [nextStep]: true },
```

Comment:
haven't tested yet.. but I would have thought that `currentStep` is completed here, not the `nextStep`?